cluster/src/main/java/org/apache/iotdb/cluster/client/ClientManager.java

```diff
@@ -0,0 +1,203 @@
+package org.apache.iotdb.cluster.client;
+
+import org.apache.iotdb.cluster.ClusterIoTDB;
+import org.apache.iotdb.cluster.client.async.AsyncDataClient;
+import org.apache.iotdb.cluster.client.async.AsyncMetaClient;
+import org.apache.iotdb.cluster.client.sync.SyncDataClient;
+import org.apache.iotdb.cluster.client.sync.SyncMetaClient;
+import org.apache.iotdb.cluster.rpc.thrift.Node;
+import org.apache.iotdb.cluster.rpc.thrift.RaftService;
+
+import com.google.common.collect.Maps;
+import org.apache.commons.pool2.KeyedObjectPool;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+/**
+ * One should borrow the reusable client from this class and return the client after use. The
+ * underlying client pool is powered by Apache Commons Pool. The class provided 3 default pool group
+ * according to current usage: ClusterClients, DataGroupClients, MetaGroupClients.
+ *
+ * <p>ClusterClients implement the data query and insert interfaces such as query and non-query
+ * request
+ *
+ * <p>DataGroupClient implement the data group raft rpc interfaces such as appendEntry,
+ * appendEntries, sendHeartbeat, etc.
+ *
+ * <p>MetaGroupClient implement the meta group raft rpc interfaces such as appendEntry,
+ * appendEntries, sendHeartbeat, etc.
+ *
+ * <p>TODO: We can refine the client structure by reorg the interfaces defined in cluster-thrift.
+ */
+public class ClientManager implements IClientPool {
+
+  private static final Logger logger = LoggerFactory.getLogger(ClusterIoTDB.class);
+
+  private Map<ClientCategory, KeyedObjectPool<Node, RaftService.AsyncClient>> asyncClientPoolMap;
+  private Map<ClientCategory, KeyedObjectPool<Node, RaftService.Client>> syncClientPoolMap;
+
+  public enum Type {
+    ClusterClient,
+    DataGroupClient,
+    MetaGroupClient
+  }
+
+  public ClientManager(boolean isAsyncMode, Type type) {
```

Comment:
That's not a good idea from my view. `ClientManager` should know nothing about `ClusterConfig` to make it reusable and write UT easily. Same for all of clients implementation, I didn't parameterize the config in clients to avoid the refactor breaking original logic too much. Ideally, all the config should pass into client via parameters.

Reply:
OK